compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java

```diff
@@ -1552,6 +1564,81 @@ private void defineMembers(List<BLangTypeDefinition> typeDefNodes, SymbolEnv pkg
         }
     }
 
+    private void validateReadOnlyIntersectionTypeDefinitions(List<BLangTypeDefinition> typeDefNodes) {
+        List<BType> loggedTypes = new ArrayList<>();
+
+        for (BLangTypeDefinition typeDefNode : typeDefNodes) {
+            BLangType typeNode = typeDefNode.typeNode;
+            NodeKind kind = typeNode.getKind();
+            if (kind == NodeKind.INTERSECTION_TYPE_NODE) {
+                BType currentType = typeNode.type;
+
+                if (currentType.tag != TypeTags.INTERSECTION) {
+                    continue;
+                }
+
+                BIntersectionType intersectionType = (BIntersectionType) currentType;
+
+                BType effectiveType = intersectionType.effectiveType;
+                if (loggedTypes.contains(effectiveType)) {
+                    continue;
+                }
+
+                loggedTypes.add(effectiveType);
```

Comment:
If the purpose of the `loggedTypes` list is to keep track of the already checked types, a Set would be a better option from a performance POV right? And the above can even be written as,
```java
if (!loggedTypes.add(effectiveType)) {
    continue;
}
```

Reply:
+1, will send a separate PR to fix this.

Reply:
Fixed in d964ea0.